native/cocos/2d/renderer/Batcher2d.h

```diff
@@ -0,0 +1,115 @@
+#pragma once
+#include <2d/renderer/UIMeshBuffer.h>
+#include <cocos/2d/renderer/RenderDrawInfo.h>
+#include <cocos/2d/renderer/UIMeshBuffer.h>
+#include <cocos/base/TypeDef.h>
+#include <cocos/core/assets/Material.h>
+#include <cocos/renderer/gfx-base/GFXTexture.h>
+#include <cocos/renderer/gfx-base/states/GFXSampler.h>
+#include <cocos/scene/DrawBatch2D.h>
+#include <unordered_map>
+#include <vector>
+#include "base/Ptr.h"
+
+namespace cc {
+class Root;
+class Batcher2d {
+public:
+    Batcher2d();
+    explicit Batcher2d(Root* root);
+    ~Batcher2d();
+
+    void syncMeshBuffersToNative(std::vector<UIMeshBuffer*>&& buffers, uint32_t length);
+
+    bool initialize();
+    void update();
+    void uploadBuffers();
+    void reset();
+    void updateVertDirtyRenderer();
+
+    void addRootNode(Node* node);
+
+public:
+    void addVertDirtyRenderer(RenderDrawInfo* entity);
+
+    UIMeshBuffer* getMeshBuffer(index_t bufferId);
+    gfx::Device* getDevice();
+
+    void updateDescriptorSet();
+
+public:
+    inline ccstd::vector<scene::DrawBatch2D*> getBatches() { return this->_batches; }
+
+    void fillBuffersAndMergeBatches();
+    void walk(Node* node);
+    void generateBatch(RenderDrawInfo* entity);
+    void resetRenderStates();
+
+private:
+    Root* _root{nullptr};
+    ccstd::vector<Node*> _rootNodeArr{};
+
+private:
+    ccstd::vector<scene::DrawBatch2D*> _batches{};
+    ccstd::vector<RenderDrawInfo*> _vertDirtyRenderers{};
+    memop::Pool<scene::DrawBatch2D> _drawBatchPool;
+
+    gfx::Device* _device{nullptr}; //use getDevice()
+
+    RenderDrawInfo* _currEntity{nullptr};
+    index_t _currBID{-1};
+    index_t _indexStart{0};
+    uint32_t _currHash{0};
+    index_t _currLayer{0};
+
+    Material* _currMaterial{nullptr};
+    gfx::Texture* _currTexture{nullptr};
+    index_t _currTextureHash{0};
+    gfx::Sampler* _currSampler{nullptr};
+    index_t _currSamplerHash{0};
+
+private:
+    ccstd::unordered_map<ccstd::hash_t, gfx::DescriptorSet*> _descriptorSetCache{};
+    gfx::DescriptorSetInfo _dsInfo{};
+    gfx::DescriptorSet* getDescriptorSet(gfx::Texture* texture, gfx::Sampler* sampler, gfx::DescriptorSetLayout* _dsLayout);
+
+private:
+    inline void fillIndexBuffers(RenderDrawInfo* entity){
```

Comment:
It is better implement it in cpp until you find it will improve performance.

Reply:
it does improve performance. The FPS improves to 50+ from 40+.